File: contest/5/main.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

int main() {

    int numTests;
    cin >> numTests;

    for (int i = 0; i < numTests; ++i) {

        int numCars, carCapacity;
        cin >> numCars >> carCapacity;

        int numBoxes;
        cin >> numBoxes;
        vector<int> boxesWeights(numBoxes);

        for (int i = 0; i < numBoxes; ++i) {
            int weight;
            cin >> weight;
            boxesWeights[i] = pow(2, weight);
        }

        sort(boxesWeights.begin(), boxesWeights.end());

        // for (int weight : boxesWeights) cout << weight << ' ';
        // cout << endl;

        int numTransportations = 0, numLoadedCars = 0, curWeight = 0;
        while (!boxesWeights.empty()) {
            if (curWeight + boxesWeights.back() <= carCapacity) {
                curWeight += boxesWeights.back();
                // cout << "Take weight = " << boxesWeights.back() << " curWeight = " << curWeight << endl;
                boxesWeights.pop_back();

                // cout << "Current boxes: ";
                for (int weight : boxesWeights) cout << weight << ' ';
                cout << endl;

            } else {
                for (int j = boxesWeights.size() - 1; j >= 0; --j) {
                    if (curWeight + boxesWeights[j] <= carCapacity) {
                        curWeight += boxesWeights[j];
                        // cout << "Take weight = " << *(boxesWeights.begin() + j) << " curWeight = " << curWeight << endl;
                        boxesWeights.erase(boxesWeights.begin() + j);
                        // cout << "Current boxes: ";
                        for (int weight : boxesWeights) cout << weight << ' ';
                        cout << endl;
                    }
                }
                ++numLoadedCars;
                // cout << "Car is loaded" << endl;
                if (numLoadedCars >= numCars) {
                    ++numTransportations;
                    // cout << "Cars going" << endl;
                    numLoadedCars = 0;
                }
                curWeight = 0;
            }
        }

        if (curWeight > 0 || numLoadedCars > 0) ++numTransportations;

        cout << numTransportations << endl;

    }

    return 0;
    
}

File: contest/5/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"math"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var numTests int
	fmt.Fscan(in, &numTests)

	for i := 0; i < numTests; i++ {
		var numCars, carCapacity int
		fmt.Fscan(in, &numCars, &carCapacity)

		var numBoxes int
		fmt.Fscan(in, &numBoxes)
		boxesWeights := make([]int, numBoxes)

		for i := 0; i < numBoxes; i++ {
			var weight int
			fmt.Fscan(in, &weight)
			boxesWeights[i] = int(math.Pow(2, float64(weight)))
		}

		sort.Ints(boxesWeights)

		numTransportations, numLoadedCars, curWeight := 0, 0, 0
		for len(boxesWeights) > 0 {
			if curWeight + boxesWeights[len(boxesWeights)-1] <= carCapacity {
				curWeight += boxesWeights[len(boxesWeights)-1]
				boxesWeights = boxesWeights[:len(boxesWeights)-1]
			} else {
				for j := len(boxesWeights) - 1; j >= 0; j-- {
					if curWeight + boxesWeights[j] <= carCapacity {
						curWeight += boxesWeights[j]
						boxesWeights = append(boxesWeights[:j], boxesWeights[j+1:]...)
					}
				}
				numLoadedCars++
				if numLoadedCars >= numCars {
					numTransportations++
					numLoadedCars = 0
				}
				curWeight = 0
			}
		}

		if curWeight > 0 || numLoadedCars > 0 {
			numTransportations++
		}

		fmt.Fprintln(out, numTransportations)
	}
}
